agent/immortalstreams/backedpipe/backed_reader_test.go

```diff
@@ -0,0 +1,612 @@
+package backedpipe_test
+
+import (
+	"context"
+	"io"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/agent/immortalstreams/backedpipe"
+	"github.com/coder/coder/v2/testutil"
+)
+
+// mockReader implements io.Reader with controllable behavior for testing
+type mockReader struct {
+	mu       sync.Mutex
+	data     []byte
+	pos      int
+	err      error
+	readFunc func([]byte) (int, error)
+}
+
+func newMockReader(data string) *mockReader {
+	return &mockReader{data: []byte(data)}
+}
+
+func (mr *mockReader) Read(p []byte) (int, error) {
+	mr.mu.Lock()
+	defer mr.mu.Unlock()
+
+	if mr.readFunc != nil {
+		return mr.readFunc(p)
+	}
+
+	if mr.err != nil {
+		return 0, mr.err
+	}
+
+	if mr.pos >= len(mr.data) {
+		return 0, io.EOF
+	}
+
+	n := copy(p, mr.data[mr.pos:])
+	mr.pos += n
+	return n, nil
+}
+
+func (mr *mockReader) setError(err error) {
+	mr.mu.Lock()
+	defer mr.mu.Unlock()
+	mr.err = err
+}
+
+func TestBackedReader_NewBackedReader(t *testing.T) {
+	t.Parallel()
+
+	errorChan := make(chan backedpipe.ErrorEvent, 1)
+	br := backedpipe.NewBackedReader(errorChan)
+	require.NotNil(t, br)
+	require.Equal(t, uint64(0), br.SequenceNum())
+	require.False(t, br.Connected())
+}
+
+func TestBackedReader_BasicReadOperation(t *testing.T) {
+	t.Parallel()
+	ctx := testutil.Context(t, testutil.WaitShort)
+
+	errorChan := make(chan backedpipe.ErrorEvent, 1)
+	br := backedpipe.NewBackedReader(errorChan)
+	reader := newMockReader("hello world")
+
+	// Connect the reader
+	seqNum := make(chan uint64, 1)
+	newR := make(chan io.Reader, 1)
+
+	go br.Reconnect(seqNum, newR)
+
+	// Get sequence number from reader
+	seq := testutil.RequireReceive(ctx, t, seqNum)
+	require.Equal(t, uint64(0), seq)
+
+	// Send new reader
+	testutil.RequireSend(ctx, t, newR, io.Reader(reader))
+
+	// Read data
+	buf := make([]byte, 5)
+	n, err := br.Read(buf)
+	require.NoError(t, err)
+	require.Equal(t, 5, n)
+	require.Equal(t, "hello", string(buf))
+	require.Equal(t, uint64(5), br.SequenceNum())
+
+	// Read more data
+	n, err = br.Read(buf)
+	require.NoError(t, err)
+	require.Equal(t, 5, n)
+	require.Equal(t, " worl", string(buf))
+	require.Equal(t, uint64(10), br.SequenceNum())
+}
+
+func TestBackedReader_ReadBlocksWhenDisconnected(t *testing.T) {
+	t.Parallel()
+	ctx := testutil.Context(t, testutil.WaitShort)
+
+	errorChan := make(chan backedpipe.ErrorEvent, 1)
+	br := backedpipe.NewBackedReader(errorChan)
+
+	// Start a read operation that should block
+	readStarted := make(chan struct{}, 1)
+	readDone := make(chan struct{})
+	var readErr error
+	var readBuf []byte
+	var readN int
+
+	go func() {
+		defer close(readDone)
+		// Signal that we're about to start the read
+		readStarted <- struct{}{}
+		buf := make([]byte, 10)
+		readN, readErr = br.Read(buf)
+		readBuf = buf[:readN]
+	}()
+
+	// Wait for read to start
+	testutil.RequireReceive(ctx, t, readStarted)
```

Comment:
`readStarted` doesn't actually verify that the `Read` call has started, since it happens on line 121 and read doesn't happen until line 123. As such it doesn't really add anything to this test and should get removed.